Restore configurable secure session timeout domain setting

orangejenny opened the pull request:
#### SUMMARY
https://trello.com/c/6IllAnGb/64-alaska-15-mins-logout-timer

This restores the configurable timer that was added in https://github.com/dimagi/commcare-hq/pull/27571 and then removed in https://github.com/dimagi/commcare-hq/pull/27725 because I thought no real projects needed it.

When this was live and being used for testing purposes, QA reported issues with the timeout length sometimes reverting to the environment default (30 min) instead of the configured length. I believe that was caused by users going between domain-specific and non-domain-specific views, with the non-domain-specific views resetting the timeout length. This PR addresses that issue by storing the timeout length in the session.

It's probably easiest to skim the commit that adds tests first and then review by commit.

##### FEATURE FLAG
Allow domain to override default length of inactivity timeout

##### RISK ASSESSMENT / QA PLAN
[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-1396)

Planning to check proxy logs after deploy.

##### PRODUCT DESCRIPTION
Same as https://github.com/dimagi/commcare-hq/pull/27571

orangejenny committed:
Added last_request to SessionDetailsView

This reapplies 0057fb56441325f4363da047a2ddb97838575e79, the last commit from
https://github.com/dimagi/commcare-hq/pull/27725, which was a bug fix.

orangejenny committed:
Reverted "Remove domain-level inactivity timeout"

This reverts https://github.com/dimagi/commcare-hq/pull/27725

orangejenny committed:
Extracted TimeoutMiddleware.update_secure_session

orangejenny committed:
Updated timeout to reflect the lower of current timeout and previous timeout

Prevent timeout from reverting to environment-level timeout when user visits a page that isn't domain-specific.

orangejenny committed:
Added warning that timeout length is sticky

orangejenny committed:
Extracted all timeout calculations into _get_timeout

orangejenny committed:
Added tests

orangejenny committed:
Removed unused imports

orangejenny committed:
Added secure_session and secure_session_timeout to ping responses

millerdev reviewed (approved):
This seems to be implementing a parallel session timeout mechanism to the default mechanism implemented by the session storage back-end. Was that an intentional design choice?

orangejenny committed:
Fixed tests

esoergel commented on corehq/apps/hqwebapp/session_details_endpoint/views.py:
```diff
@@ -66,8 +67,7 @@ def post(self, request, *args, **kwargs):
 
-        session.set_expiry(timeout * 60)
-        session['last_request'] = json_format_datetime(datetime.datetime.utcnow())
+        TimeoutMiddleware.update_secure_session(session, secure_session, timeout)
```
:heart:

esoergel commented on corehq/apps/hqwebapp/session_details_endpoint/views.py:
```diff
@@ -57,17 +57,7 @@ def post(self, request, *args, **kwargs):
-                if domain_obj:
-                    timeout = domain_obj.secure_timeout or timeout
-        else:
-            timeout = settings.INACTIVITY_TIMEOUT
```
nice, was gonna comment about this

esoergel commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -500,19 +500,21 @@ def logout(req, default_domain_redirect='domain_login'):
-        'last_request': request.session.get('last_request'),
-        'username': request.user.username,
-    })
+    return _ping_response(request)
```
Since these two views are identical, you could also keep just one and maintain two different entries in `urls.py`, both pointing to that.

orangejenny replied:
good point, 5bfbb52f8cc

esoergel commented on corehq/middleware.py:
```diff
@@ -113,11 +113,30 @@ def process_response(self, request, response):
 class TimeoutMiddleware(MiddlewareMixin):
 
+    @classmethod
+    def update_secure_session(cls, session, is_secure, domain=None):
```
it looks like this method does two things, depending on `is_secure` - bumps the timeout, and toggles secure sessions.  I think it'd be clearer to separate the two.

orangejenny replied:
I tried this out but didn't especially like the result. I like that they're coupled because you shouldn't be calling them separately.

esoergel reviewed (approved):
no blocking comments

orangejenny committed:
Combined ping_login and ping_session

orangejenny committed:
Merge branch 'master' into revert-27725-jls/kill-configurable-timeout

orangejenny committed:
Replaced last_request with session_expiry

millerdev commented on corehq/middleware.py:
```diff
@@ -116,10 +116,10 @@ class TimeoutMiddleware(MiddlewareMixin):
         timeout = cls._get_timeout(session, is_secure, domain)
         session['secure_session_timeout'] = timeout
         session.set_expiry(timeout * 60)
+        session['session_expiry'] = session.get_expiry_date()
```
Curious why this needs to be stored in the session? Would it work just as well to call `session.get_expiry_date()` in `ping_response`?

orangejenny replied:
Hmm...maybe it doesn't need to be stored here. I'll keep that in mind (I'm still working on this, adding in the logic we discussed to use the minimum timeout length of all domains where the user is a member).

orangejenny replied:
After working with this code some more, I'm inclined to keep this as-is. The javascript now uses `session_expiry`, and storing it in the session makes it easily available on page load.

orangejenny committed:
Merge branch 'master' into revert-27725-jls/kill-configurable-timeout

orangejenny committed:
Added session expiry date to js

This makes a few minor fixes:
- Pass the session expiry to js on page load
- If expiry isn't available (which shouldn't happen) wait 10 minutes to ping, not 1.
Waiting only 1 minute makes the 2-minute warning modal show up.
- Enable this UI whenever the user's session is secure. Previously you could get logged out
but not see the popup if you were viewing a non-secure domain but had a secure session (likely from
having just visited a secure domain).

orangejenny committed:
Updated logic to use minimum timeout from all of user's domains

orangejenny commented:
I made a couple of fixes and (hopefully) simplifications. I'm still working on additional test coverage and will ping for re-review when that's finished.

orangejenny commented on corehq/apps/hqwebapp/session_details_endpoint/tests.py:
```diff
@@ -69,6 +71,59 @@ def test_session_details_view_updates_session(self):
+        self.client.get(reverse('domain_homepage', args=[self.domain.name]))
+        self.assertEqual(False, self.client.session.get('secure_session'))
+        self.assertEqual(settings.INACTIVITY_TIMEOUT, self.client.session.get('secure_session_timeout'))
+
```
@millerdev and I also discussed this and settled on using the minimum timeout for all domains where the user is a member, because this matches the spirit of the secure_session flag (where you get the "most secure" experience or all of the domains you belong to). https://github.com/dimagi/commcare-hq/pull/27761/commits/07422d680bf9892a02dd1f96cd606cd3b1eacaf6 implements this. Definitely agree on test coverage, still working on that.

orangejenny replied:
015d8a524a6839d7e3fb273b1e9479a41de3db76 adds tests

orangejenny committed:
Updated tests

orangejenny committed:
Added check for request.project back to base.html

I switched this earlier to check for request.secure_session, so that the popups
can show in domains that aren't secure (necessary if the user is a member of a
different, secure domain). But the domain check is still necessary, because the
login page used in the popup is the domain-specific login.

orangejenny committed:
Updated timeout length fallback logic

Only use the default secure timeout if no other timeouts are available
(which should never happen). Previously, it wasn't possible for a domain
to configure a timeout longer than the default, because the min call
included the default.

orangejenny committed:
Added more tests

orangejenny committed:
Added docs

orangejenny commented:
Added more tests and docs; this is ready for re-review now. @esoergel @millerdev

orangejenny committed:
Fixed typos

orangejenny committed:
Fixed time zone for javascript

The session expiry date is in UTC, but JavaScript dates use the user's time zone.
4671065 included time zone information for the initial expiry date, but calls to
ping_login were not getting the time zone, so js was misunderstanding the expiry.

orangejenny committed:
Updated log to show expiry date in same time zone as user

orangejenny committed:
Fixed bug with unknown expiry immediately after re-logging in

esoergel reviewed (approved):
Great tests!  Nothing blocking

orangejenny committed:
Update docs/middleware.rst

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

orangejenny committed:
Switched from hasattr to in

orangejenny committed:
Merge branch 'revert-27725-jls/kill-configurable-timeout' of github.com:dimagi/commcare-hq into revert-27725-jls/kill-configurable-timeout

orangejenny committed:
Renamed domain => insecure_domain

orangejenny committed:
Update corehq/apps/hqwebapp/tests/test_timeout_middleware.py

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

orangejenny commented on corehq/apps/hqwebapp/tests/test_timeout_middleware.py:
```diff
@@ -0,0 +1,106 @@
+        self.user = WebUser.get_by_username(self.user.username)
+        self.user.add_as_web_user(self.secure_domain1.name, 'admin')
+
+        self._get_page(self.domain)
```
I like that - afc9f82c00a

orangejenny committed:
Merge branch 'revert-27725-jls/kill-configurable-timeout' of github.com:dimagi/commcare-hq into revert-27725-jls/kill-configurable-timeout

orangejenny committed:
Merge branch 'master' into revert-27725-jls/kill-configurable-timeout

orangejenny committed:
Added time zone in test
